lib/widgets/compose_box.dart

```diff
@@ -44,21 +47,75 @@ class TopicTextEditingController extends TextEditingController {
 
 enum ContentValidationError {
   empty,
-  tooLong;
+  tooLong,
+  uploadInProgress;
 
-  // Later: upload in progress; quote-and-reply in progress
+  // Later: quote-and-reply in progress
 
   String message() {
     switch (this) {
       case ContentValidationError.tooLong:
         return "Message length shouldn't be greater than 10000 characters.";
       case ContentValidationError.empty:
         return 'You have nothing to send!';
+      case ContentValidationError.uploadInProgress:
+        return 'Please wait for the upload to complete.';
     }
   }
 }
 
 class ContentTextEditingController extends TextEditingController {
+  int _uploadCounter = 0;
+
+  final Map<int, ({String filename, String placeholder})> _uploads = {};
+
+  /// The cursor position, if selection is valid and collapsed, else text.length.
+  TextRange _cursorPositionOrEnd() {
+    final TextRange selection = value.selection;
+    final String text = value.text;
+    final int index = (selection.isValid && selection.end < text.length && selection.isCollapsed)
+      ? selection.end
+      : text.length;
+    return TextRange.collapsed(index);
+  }
+
+  /// Tells the controller that a file upload has started.
+  ///
+  /// Returns an int "tag" that should be passed to registerUploadEnd on the
+  /// upload's success or failure.
+  int registerUploadStart(String filename) {
+    final tag = _uploadCounter;
+    _uploadCounter += 1;
+    final placeholder = '[Uploading $filename...]()\n\n'; // TODO(i18n)
+    _uploads[tag] = (filename: filename, placeholder: placeholder);
+    notifyListeners(); // _uploads change could affect validationErrors
+    value = value.replaced(_cursorPositionOrEnd(), placeholder);
+    return tag;
+  }
+
+  /// Tells the controller that a file upload has ended, with success or error.
+  ///
+  /// To indicate success, pass the URL to be used for the Markdown link.
+  /// Without that, failure is assumed.
+  void registerUploadEnd(int tag, Uri? url) {
+    final val = _uploads[tag];
+    assert(val != null, 'registerUploadEnd called twice for same tag');
+    final (filename: filename, placeholder: placeholder) = val!;
+    final int startIndex = text.indexOf(placeholder);
+    final replacementRange = startIndex >= 0
+      ? TextRange(start: startIndex, end: startIndex + placeholder.length)
+      : _cursorPositionOrEnd();
+
+    // TODO(i18n)
+    value = value.replaced(
+      replacementRange,
+      url == null
+        ? '[Failed to upload file: $filename]()\n\n'
```

Comment:
```suggestion
        ? '[Failed to upload file: $filename]()\n\n' // TODO(i18n)
```

This way the comment is a lot more self-explanatory than when it's several lines above.

(This is one of the payoffs of not using something like Prettier.)

Reply:
I sort-of regularly fight Prettier about putting comments on the branches of the conditional operator (`b` and `c` in `a ? b : c`), but I think not in a case like this where the comment is short enough to share a line with one of the branches.

Incidentally this bit of code can be parsed as JavaScript, so one can conveniently see what Prettier would do with similar code:

```js
value = value.replaced(
  replacementRange,
  url == null
    ? "[Failed to upload file: $filename]()\n\n" // TODO(i18n)
    : "[$filename](${url.toString()})\n\n"
);
```

(Note the `TODO` shares a line with the "Failed to…" branch, as in your suggestion for the Dart code.)

https://prettier.io/playground/#N4Igxg9gdgLgprEAuEACdqBuBDANgVzlQF4s9CA6AJzgAddsw4ATACgB0oMMb7G4AtghgAlbFADmcADSdu6fFVwlSUfLlxz56APyoA5AG0AYtgCWuFqhgRU+ehGzNUAMwtwkqACRvLUbEIAuqwAlOyc4VD6qAD0MagAKgDyACJJrGYAjAAcUGFc2qieRj7u-kGsXsCKuBQ2AMowVGaSoQC+YRGc+iEA3CDSIBC0MGbQAM7IoNhUVBAA7gAKMwiTKHjz2ACek4MARlSMANZwMPUBcAAyLXDILnjjMiAHx6f1tIwtEshNhIOCexYzBYl3EEnw2CkxggVAE2Bgo0kyBA2HwNgGIAAFjABLgAOqYszwcYfJj1VZEsyYIlbZFgca7EAtR5UGCLQ4SOF3B5PABW4wAHvUvpYAIr4CDwbm4R6DD5UFnIvbYQG4DG0ZqwPFmZgwTHIbIABjlc0eeMOtGRGrgLMwt0GAEcJfB2cM1ijxgBaKBwFgsDE0J1mGjsyFcpD3GVPR4CMw-Kh-EDjEVwACCCOaezRcEWcCo1x90tlSZT4sltwjPMGMBV2t1+qQACZq4cLF8AMIQAThkA2gCsGPwjwSKrWkeLmEIAEkoMDYPUwM0RqnZ40tpYi3A2m0gA

Reply:
I’m not going to stand for this libel—Prettier is just fine with any or all of these comment locations:

```js
value = value.replaced(
  replacementRange,
  // 1
  url == null // 2
    ? // 3
      "[Failed to upload file: $filename]()\n\n" // 4
    : // 5
      "[$filename](${url.toString()})\n\n" // 6
  // 7
);
```

and so is `flutter format`, with minor whitespace variations!

```dart
value = value.replaced(
    replacementRange,
    // 1
    url == null // 2
        ? // 3
        '[Failed to upload file: $filename]()\n\n' // 4
        : // 5
        '[$filename](${url.toString()})\n\n' // 6
    // 7
    );
```

Reply:
> I’m not going to stand for this libel

Perhaps your comments crossed? I think Chris has already handled the subject.


> what Prettier would do with similar code:

Ah cool, that's convenient, then. A case we've used regularly in zulip-flutter that I know I've run into many times where Prettier doesn't accept a convenient comment location is on the line of an if-condition. We've written this:

```dart
if (context.mounted) {} // https://github.com/dart-lang/linter/issues/4007
else {
  return;
}
```

Prettier would turn that unhelpfully into this:
```js
if (context.mounted) {
} // https://github.com/dart-lang/linter/issues/4007
else {
  return;
}
```

So I guess I was thinking of cases like that one.